File: apps/ios/Packages/Sources/ShadowRoomTimelineFeature/RoomTimelineModels.swift
import Foundation

public struct RoomTimelineSnapshotViewState: Equatable, Sendable {
    public let roomId: String
    public let roomTitle: String?
    public let securityState: RoomTimelineSecurityState
    public let items: [RoomTimelineItemViewState]

    public init(
        roomId: String,
        roomTitle: String?,
        securityState: RoomTimelineSecurityState = .unknown,
        items: [RoomTimelineItemViewState]
    ) {
        self.roomId = roomId
        self.roomTitle = roomTitle
        self.securityState = securityState
        self.items = items
    }
}

public enum RoomTimelineSecurityState: Equatable, Sendable {
    case encrypted
    case unencrypted
    case unknown
}

public struct RoomTimelineItemViewState: Equatable, Identifiable, Sendable {
    public let messageId: String
    public let senderDisplayName: String?
    public let body: String
    public let sentAtLabel: String
    public let direction: RoomTimelineMessageDirection
    public let deliveryState: RoomTimelineDeliveryState

    public var id: String { messageId }

    public init(
        messageId: String,
        senderDisplayName: String?,
        body: String,
        sentAtLabel: String,
        direction: RoomTimelineMessageDirection,
        deliveryState: RoomTimelineDeliveryState
    ) {
        self.messageId = messageId
        self.senderDisplayName = senderDisplayName
        self.body = body
        self.sentAtLabel = sentAtLabel
        self.direction = direction
        self.deliveryState = deliveryState
    }
}

public enum RoomTimelineMessageDirection: Equatable, Sendable {
    case incoming
    case outgoing
}

public enum RoomTimelineDeliveryState: Equatable, Sendable {
    case sending
    case sent
    case delivered
    case read
    case failed
}

public enum RoomTimelineState: Equatable, Sendable {
    case loading
    case loaded(roomTitle: String?, items: [RoomTimelineItemViewState])
    case empty(roomTitle: String?)
    case failed
}

public enum RoomTimelineEvent: Equatable, Sendable {
    case refreshRequested
    case retryRequested
    case draftChanged(String)
    case sendRequested
}

File: apps/android/features/timeline/src/main/java/com/shadowchat/features/timeline/RoomTimelineModels.kt
package com.shadowchat.features.timeline

data class RoomTimelineSnapshotUi(
    val roomId: String,
    val roomTitle: String?,
    val items: List<RoomTimelineItemUi>,
)

data class RoomTimelineItemUi(
    val messageId: String,
    val senderDisplayName: String?,
    val body: String,
    val sentAtLabel: String,
    val direction: RoomTimelineMessageDirection,
    val deliveryState: RoomTimelineDeliveryState,
)

enum class RoomTimelineMessageDirection {
    Incoming,
    Outgoing,
}

enum class RoomTimelineDeliveryState {
    Sending,
    Sent,
    Delivered,
    Read,
    Failed,
}

sealed interface RoomTimelineUiState {
    data object Loading : RoomTimelineUiState
    data class Loaded(
        val roomTitle: String?,
        val items: List<RoomTimelineItemUi>,
    ) : RoomTimelineUiState
    data class Empty(val roomTitle: String?) : RoomTimelineUiState
    data object Error : RoomTimelineUiState
}

sealed interface RoomTimelineEvent {
    data object Appeared : RoomTimelineEvent
    data object RefreshRequested : RoomTimelineEvent
    data object RetryRequested : RoomTimelineEvent
}
